### ISSUE TITLE
False positive on the `unreachable_pub` lint

### ISSUE THREAD

**Issue by alexcrichton** [rust-lang/rust#52664]
The following code:

```rust
#![crate_type = "lib"]
#![feature(rust_2018_preview)]
#![warn(rust_2018_idioms)]

pub use self::m1::foo;

mod m1 {
    pub use self::m2::foo;


    mod m2 {
        pub fn foo() {}
    }
}
```

[when compiled](https://play.rust-lang.org/?gist=ff9b9d894fda08183165bd3a8f2c3b01&version=nightly&mode=release&edition=2015) yields:

```
warning: unreachable `pub` item
 --> src/lib.rs:8:5
  |
8 |     pub use self::m2::foo;
  |     ---^^^^^^^^^^^^^^^^^^^
  |     |
  |     help: consider restricting its visibility: `crate`
  |
note: lint level defined here
 --> src/lib.rs:3:9
  |
3 | #![warn(rust_2018_idioms)]
  |         ^^^^^^^^^^^^^^^^
  = note: #[warn(unreachable_pub)] implied by #[warn(rust_2018_idioms)]
  = help: or consider exporting it for use by other crates
```

but the corrected version doesn't compile!

**Comment by alexcrichton** [comment]
cc @zackmdavis, do you know where this might be coming up?

**Comment by zackmdavis** [comment]
Duplicate of #47816? I'm pretty busy right now, but might be able to take a look Saturday.

**Comment by alexcrichton** [comment]
Indeed, thanks! No rush as well :)
